### ISSUE TITLE
Why is test_scene_as_chunks running on input instead of target and what is the definition of Valid Pixels and 3D reconstruction loss?

### ISSUE THREAD

**Issue by ferbrjan** [angeladai/spsg#16]
Good afternoon,

After successfully running the testing of SPSG on our data, I was studying the code of test_scene_as_chunks.py and came across a weird part of the code. When running the trained model spsg.pth using the model() function, one of the arguments is chunk_input. From what I understand, this chunk_input variable refers to the incomplete version of the scan, while the testing part should run on the complete scan (I assume that is chunk_target_sdf variable). See screenshot below: (line 138 in test_scene_as_chunks.py

<img width="1099" alt="Snímek obrazovky 2022-08-16 v 16 56 01" src="https://user-images.githubusercontent.com/74875970/184913408-9f8c3aef-df1d-4184-91b9-8ee16aa9a42b.png">
)

To find more information about this, I was rereading your paper to understand the process entirely, and I came across two more things that are somewhat unclear to me. Can you please specify what is meant by valid pixels in section 4.2? 

<img width="446" alt="Snímek obrazovky 2022-08-16 v 17 08 23" src="https://user-images.githubusercontent.com/74875970/184914771-95f99c71-9b51-43df-bea6-258be7d6a6a4.png">

Another unclear thing is the 3D geometric reconstruction loss. Can you please define what is meant by this term?

<img width="423" alt="Snímek obrazovky 2022-08-16 v 17 10 11" src="https://user-images.githubusercontent.com/74875970/184915033-1091f483-99cb-41d6-aa26-1624c3fb82c1.png">

Thank you very much for your help,

Regards,

Jan

**Comment by wtiandong** [comment]
> Good afternoon,
> 
> After successfully running the testing of SPSG on our data, I was studying the code of test_scene_as_chunks.py and came across a weird part of the code. When running the trained model spsg.pth using the model() function, one of the arguments is chunk_input. From what I understand, this chunk_input variable refers to the incomplete version of the scan, while the testing part should run on the complete scan (I assume that is chunk_target_sdf variable). See screenshot below: (line 138 in test_scene_as_chunks.py
> 
> <img alt="Snímek obrazovky 2022-08-16 v 16 56 01" width="1099" src="https://user-images.githubusercontent.com/74875970/184913408-9f8c3aef-df1d-4184-91b9-8ee16aa9a42b.png">
> 
> )
> To find more information about this, I was rereading your paper to understand the process entirely, and I came across two more things that are somewhat unclear to me. Can you please specify what is meant by valid pixels in section 4.2?
> 
> <img alt="Snímek obrazovky 2022-08-16 v 17 08 23" width="446" src="https://user-images.githubusercontent.com/74875970/184914771-95f99c71-9b51-43df-bea6-258be7d6a6a4.png">
> 
> Another unclear thing is the 3D geometric reconstruction loss. Can you please define what is meant by this term?
> 
> <img alt="Snímek obrazovky 2022-08-16 v 17 10 11" width="423" src="https://user-images.githubusercontent.com/74875970/184915033-1091f483-99cb-41d6-aa26-1624c3fb82c1.png">
> 
> Thank you very much for your help,
> 
> Regards,
> 
> Jan

Since the authors seldom reply to issues on Github, I would like to try to answer your questions.

First, "valid pixels". When you render a TSDF volume at a specific camera pose, not all pixels on the rendered image are filled. At some locations, there isn't "mesh"  (i.e, empty), but the camera "sees" these locations. So the empty locations rendered on the image are invalid pixels. See [here](https://github.com/angeladai/spsg/blob/8d0dff65480116799cf35499a7f672c2360bec60/torch/train.py#L448)

Second, "3D geometric reconstruction loss". This is just loss = |TSDF_pred - TSDF_target|.
See [here](https://github.com/angeladai/spsg/blob/8d0dff65480116799cf35499a7f672c2360bec60/torch/train.py#L353)

**Comment by ferbrjan** [comment]
> > Good afternoon,
> > After successfully running the testing of SPSG on our data, I was studying the code of test_scene_as_chunks.py and came across a weird part of the code. When running the trained model spsg.pth using the model() function, one of the arguments is chunk_input. From what I understand, this chunk_input variable refers to the incomplete version of the scan, while the testing part should run on the complete scan (I assume that is chunk_target_sdf variable). See screenshot below: (line 138 in test_scene_as_chunks.py
> > <img alt="Snímek obrazovky 2022-08-16 v 16 56 01" width="1099" src="https://user-images.githubusercontent.com/74875970/184913408-9f8c3aef-df1d-4184-91b9-8ee16aa9a42b.png">
> > )
> > To find more information about this, I was rereading your paper to understand the process entirely, and I came across two more things that are somewhat unclear to me. Can you please specify what is meant by valid pixels in section 4.2?
> > <img alt="Snímek obrazovky 2022-08-16 v 17 08 23" width="446" src="https://user-images.githubusercontent.com/74875970/184914771-95f99c71-9b51-43df-bea6-258be7d6a6a4.png">
> > Another unclear thing is the 3D geometric reconstruction loss. Can you please define what is meant by this term?
> > <img alt="Snímek obrazovky 2022-08-16 v 17 10 11" width="423" src="https://user-images.githubusercontent.com/74875970/184915033-1091f483-99cb-41d6-aa26-1624c3fb82c1.png">
> > Thank you very much for your help,
> > Regards,
> > Jan
> 
> Since the authors seldom reply to issues on Github, I would like to try to answer your questions.
> 
> First, "valid pixels". When you render a TSDF volume at a specific camera pose, not all pixels on the rendered image are filled. At some locations, there isn't "mesh" (i.e, empty), but the camera "sees" these locations. So the empty locations rendered on the image are invalid pixels. See [here](https://github.com/angeladai/spsg/blob/8d0dff65480116799cf35499a7f672c2360bec60/torch/train.py#L448)
> 
> Second, "3D geometric reconstruction loss". This is just loss = |TSDF_pred - TSDF_target|. See [here](https://github.com/angeladai/spsg/blob/8d0dff65480116799cf35499a7f672c2360bec60/torch/train.py#L353)

Perfect! This helps a lot! 

Now the only missing part is the usage of chunk_input. Hopefully, someone will clarify that.

**Comment by wtiandong** [comment]
BTW, may I ask what gan loss you use and how many epochs you train the network when reproducing SPSG code? I notice that training is very slow when using vanilla gan loss.

> > > Good afternoon,
> > > After successfully running the testing of SPSG on our data, I was studying the code of test_scene_as_chunks.py and came across a weird part of the code. When running the trained model spsg.pth using the model() function, one of the arguments is chunk_input. From what I understand, this chunk_input variable refers to the incomplete version of the scan, while the testing part should run on the complete scan (I assume that is chunk_target_sdf variable). See screenshot below: (line 138 in test_scene_as_chunks.py
> > > <img alt="Snímek obrazovky 2022-08-16 v 16 56 01" width="1099" src="https://user-images.githubusercontent.com/74875970/184913408-9f8c3aef-df1d-4184-91b9-8ee16aa9a42b.png">
> > > )
> > > To find more information about this, I was rereading your paper to understand the process entirely, and I came across two more things that are somewhat unclear to me. Can you please specify what is meant by valid pixels in section 4.2?
> > > <img alt="Snímek obrazovky 2022-08-16 v 17 08 23" width="446" src="https://user-images.githubusercontent.com/74875970/184914771-95f99c71-9b51-43df-bea6-258be7d6a6a4.png">
> > > Another unclear thing is the 3D geometric reconstruction loss. Can you please define what is meant by this term?
> > > <img alt="Snímek obrazovky 2022-08-16 v 17 10 11" width="423" src="https://user-images.githubusercontent.com/74875970/184915033-1091f483-99cb-41d6-aa26-1624c3fb82c1.png">
> > > Thank you very much for your help,
> > > Regards,
> > > Jan
> > 
> > 
> > Since the authors seldom reply to issues on Github, I would like to try to answer your questions.
> > First, "valid pixels". When you render a TSDF volume at a specific camera pose, not all pixels on the rendered image are filled. At some locations, there isn't "mesh" (i.e, empty), but the camera "sees" these locations. So the empty locations rendered on the image are invalid pixels. See [here](https://github.com/angeladai/spsg/blob/8d0dff65480116799cf35499a7f672c2360bec60/torch/train.py#L448)
> > Second, "3D geometric reconstruction loss". This is just loss = |TSDF_pred - TSDF_target|. See [here](https://github.com/angeladai/spsg/blob/8d0dff65480116799cf35499a7f672c2360bec60/torch/train.py#L353)
> 
> Perfect! This helps a lot!
> 
> Now the only missing part is the usage of chunk_input. Hopefully, someone will clarify that.

**Comment by ferbrjan** [comment]
I used the default setting from the example command. It was slow but the trained model was working faster than the one provided by the authors.

sorry but I have not experimented with the training too much. :-(
